File: src/main/java/mvp/user/UserFragment.java
package mvp.user;

import java.util.List;

import chapter2OOP.User;
import mvp.BaseFragment;
import mvp.UserContract;

public class UserFragment extends BaseFragment<UserContract.UserView, UserContract.UserPresenter>
        implements UserContract.UserView {

    @Override
    public void onCreate() {
        super.onCreate();
        presenter = InitUserFragment.initUserModule();
    }

    @Override
    public void onCreateView() {
        super.onCreateView();
        presenter.fetchUsers();
    }

    @Override
    public void displayUser(List<User> users) {
        for (int i = 0; i < users.size(); i++) {

            User user = users.get(i);
            System.out.println("UserFragment displayUser() USER: " + user.getFullName());
        }

    }

    @Override
    public void displayError(String message) {

    }

    static class InitUserFragment {

        public static UserContract.UserPresenter initUserModule() {
            UserContract.UserInteractor userInteractor = new UserInteractor();
            return new UserPresenter(userInteractor);

        }

    }
}

File: tutorial/src/main/java/com/smarttoolfactory/tutorial/chapter2OOP/mvp/user/UserFragment.kt
package com.smarttoolfactory.tutorial.chapter2OOP.mvp.user



import com.smarttoolfactory.tutorial.chapter2OOP.User
import com.smarttoolfactory.tutorial.chapter2OOP.mvp.BaseFragment
import com.smarttoolfactory.tutorial.chapter2OOP.mvp.BaseInteractor
import com.smarttoolfactory.tutorial.chapter2OOP.mvp.BasePresenter
import com.smarttoolfactory.tutorial.chapter2OOP.mvp.UserContract
import java.util.*

class UserFragment : BaseFragment<UserContract.UserView, UserContract.UserPresenter>(),
    UserContract.UserView {

    init {
        presenter = InitUserFragment.initUserModule()
    }

    override fun onCreate() {
        super.onCreate()
        println("😍 UserFragment onCreate()")
    }

    override fun onCreateView() {
        super.onCreateView()
        presenter?.fetchUsers()
    }

    override fun displayUser(users: List<User>) {
        for (i in users.indices) {
            val user = users[i]
            println("🔥🔥 UserFragment displayUser() USER: " + user.fullName)
        }
    }

    override fun displayError(message: String?) {}

    internal object InitUserFragment {
        fun initUserModule(): UserContract.UserPresenter {
            val userInteractor: UserContract.UserInteractor = UserInteractor()
            return UserPresenter(userInteractor)
        }
    }
}


class UserPresenter(interactor: UserContract.UserInteractor) :
    BasePresenter<UserContract.UserView, UserContract.UserInteractor>(interactor), UserContract.UserPresenter,
    UserContract.UserOutput {

    override fun fetchUsers() {
        interactor?.fetchUsersFromNetwork()
    }

    override fun onUsersFetched(users: List<User>) {
        view?.displayUser(users)
    }

    override fun onErrorOccurred(message: String) {
        view?.displayError(message)
    }
}

class UserInteractor : BaseInteractor<UserContract.UserOutput>(), UserContract.UserInteractor {

    override fun fetchUsersFromNetwork() {
        val userList = GetUsers.users
        output?.onUsersFetched(userList)
        println("🤩 UserInteractor fetchUsersFromNetwork() userList: $userList")
    }

    internal object GetUsers {
        val users: List<User>
            get() {
                val userList: MutableList<User> = ArrayList()
                val user = User("John", "Smith")
                userList.add(user)
                return userList
            }
    }
}